BuildIndex.py

```diff
@@ -20,40 +20,81 @@
+
+    # Retrieve the change in percentage, value
     try:
-        if DEBUG_MODE: log("About to scrape yahoo for change")
-        change_amt, change_pct = scrape_yahoo_change(symbol)
+        action = "change amounts"
+        if DEBUG_MODE: log("Retrieving {} for {}".format(action, symbol))
+        change_amt, change_pct = scrape_yahoo_change(content)
         log("[{}] Retrieved change amount: {}\tchange percent: {}".format(symbol, change_amt, change_pct))
+    except Exception as e:
+        print(e)
+        log("Error retrieving {} data for: {}".format(action, symbol), e)
 
-        if DEBUG_MODE: log("About to scrape yahoo for price")
-        price = scrape_yahoo_price(symbol)
+    # Retrieve the Price amount
+    try:
+        action = "price"
+        if DEBUG_MODE: log("Retrieving {} for {}".format(action, symbol))
+        price = scrape_yahoo_price(content)
         log("[{}] Retrieved price: {}".format(symbol, price))
+    except Exception as e:
+        print(e)
+        log("Error retrieving {} data for: {}".format(action, symbol), e)
 
-        if DEBUG_MODE: log("About to scrape yahoo for mkt_cap")
-        market_cap = scrape_yahoo_mkt_cap(symbol)
+    # Retrieve market cap
+    try:
+        action = "market cap"
+        if DEBUG_MODE: log("Retrieving {} for {}".format(action, symbol))
+        market_cap = scrape_yahoo_mkt_cap(content)
         log("[{}] Retrieved mkt cap: {}".format(symbol, market_cap))
+    except Exception as e:
+        print(e)
+        log("Error retrieving {} data for: {}".format(action, symbol), e)
 
-        if DEBUG_MODE: log("About to scrape yahoo for name")
-        name = scrape_yahoo_name(symbol)
+    # Retrieve the name
+    try:
+        action = "name"
+        if DEBUG_MODE: log("Retrieving {} for {}".format(action, symbol))
+        name = scrape_yahoo_name(symbol, content)
         log("[{}] Retrieved name: {}".format(symbol, name))
-        output = {
-            "price": price,
-            "refresh_time": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
-            "market_cap": market_cap,
-            "name": name,
-            "change_from_open": change_amt,
-            "change_from_open_percent": change_pct
-        }
-        return output
     except Exception as e:
         print(e)
-        log("Error trying to scrape data for: {}".format(symbol), e)
-        return None
+        log("Error retrieving {} data for: {}".format(action, symbol), e)
+
+    # Build the new dictionary using the data we just retrieved
+    new_data = {
+        "price": price,
+        "refresh_time": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
+        "market_cap": market_cap,
+        "name": name,
+        "change_from_open": change_amt,
+        "change_from_open_percent": change_pct
+    }
+
+    # Check our new dictionary. If we have any blanks, use the old dictionary if that's
+    # not blank
+    for key in new_data.keys():
```

Comment:
This is intended to serve as a fallback for failed retrievals - if the data is already present in the json file that gets written at the end of each run, we can use that when not available from scraping